Dont hold onto references of saved tensors in backward

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88248
* __->__ #88247
* #88208

This improves memory compression of resnet18 on inductor non-cudagraphs from .78 -> .0.84.

eellison committed:
Dont hold onto references of saved tensors in backward

[ghstack-poisoned]

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -472,12 +472,16 @@ def forward(ctx, *deduped_flat_tensor_args):
                 with context(), track_graph_compiling("backward", True):
                     CompiledFunction.compiled_bw = aot_config.bw_compiler(
-                        bw_module, all_args
+                        bw_module, fake_args
```
@Chillee how come the backward compiler needs fake args, didn't we already synthesize then when creating the joint graph?

Chillee replied:
What do you mean? We created them, but in the future, it'd be possible for them to be other tensor subclasses.

Also, the other issue is that need to pass symints from the backwards to the forwards pass.

ezyang replied:
It's hard for me to believe this case actually works correctly lol.

ezyang replied:
Like, how do you even fakify an arbitrary torch dispatch tensor subclass

ezyang reviewed (approved):
Nice catch. Any chance we can test it?

eellison commented:
Filed https://github.com/pytorch/torchdynamo/issues/1829: at the moment this is pretty difficult to test.

eellison committed:
Update on "Dont hold onto references of saved tensors in backward"


This improves memory compression of resnet18 on inductor non-cudagraphs from .78 -> .0.84.


[ghstack-poisoned]

eellison committed:
Update on "Dont hold onto references of saved tensors in backward"


This improves memory compression of resnet18 on inductor non-cudagraphs from .78 -> .0.84.


[ghstack-poisoned]

eellison commented:
@pytorchbot merge -f "flakey rocm test"

eellison commented on functorch/_src/aot_autograd.py:
```diff
@@ -472,12 +472,15 @@ def forward(ctx, *deduped_flat_tensor_args):
             all_args = list(ctx.symints) + list(ctx.saved_tensors) + list(contiguous_args)
+            del contiguous_args
             if CompiledFunction.compiled_bw is None:
+                # TODO - pass in fake tensors ?
```
this wasn't actually necessary for memory savings and had a bad interaction with the pr above this + didn't repro with minifier - deferring on passing in fake for now
